enhancements/update/reduced_reboots.md

```diff
@@ -0,0 +1,328 @@
+and the intersection of the specific features of OpenShift. We should work to
+define and implement upstream testing changes which give us an appropriate level
+of confidence that N-2 version skew issues would be identified in the community
+whenever possible.
+
+#### OTA - Implement Downstream Paused Worker Pool Upgrade Tests
+
+In parallel with efforts to revamp upstream version skew testing we must also
+implement downstream version skew testing which includes any additional tests
+required for OpenShift specific implementation details.
+
+We will achieve this by delivering upgrade jobs which pause the Worker MachineConfigPool
+then upgrade from 4.x to 4.x+1 to 4.x+2. We will run conformance tests from 4.x
+after upgrading to 4.x+1 in order to ensure that we continue to provide a baseline
+feature set, then again after upgrading to 4.x+2, finally after unpausing the
+Worker MCP we will run 4.x+2 tests.
+
+Given the complexity of these test jobs we should expect that they may take
+longer than the current four hour limit for test jobs. Rather than compromising
+on test completeness we will seek to extend test duration limits or find other
+ways to meet these testing demands.
+
+#### Teams with Host Components - Allow N-2 Host Component Version Skew
+
+All teams which own components that directly interface with or ship host based
+components will need to ensure that they're broadening their compatibility to
+allow for N-2 version skew between host and cluster scoped resources.
+
+This would include for example the SDN DaemonSets in 4.10 remaining compatible
+with OVS and any other host components in 4.10, 4.9, and 4.8. On a case by case
+basis teams should decide whether it makes more sense to maintain a broader
+compatibility matrix or that N-1 bits and MachineConfig are backported to N-2 and
+upgrade graph is amended with these new minimum version requirements.
+
+For instance, if 4.9.12 is the minimum version for 4.9 to 4.10 upgrades we'd
+ensure that the next 4.8.z shipping after 4.9.12 has RHCOS bits and MachineConfig
+which offer parity with 4.9.12 so that it's not required that we reboot into
+4.9.12. If teams choose to pursue this option they will need to continue to ensure
+that 4.7 to 4.8.z and 4.8.z-n to 4.8.z upgrades continue to work as well.
+
+Teams which believe this is not achievable or the level of effort is extremely
+high should document those findings.
+
+Thus far RHCOS, Node, MCO, SDN, Containers, and PSAP teams are known to fall into
+this group of teams which have components coupled to the host.
+
+#### MCO - Widen Node Constraints to allow for N-2
+
+Building upon the EUS-to-EUS upgrade MVP work(https://github.com/openshift/enhancements/blob/master/enhancements/update/eus-upgrades-mvp.md#mco---enforce-openshifts-defined-host-component-version-skew-policies)
+to allow MCO to enforce host constraints we will broaden those constraints to
+enable host component version skew.
+
```

Comment:
Another constraint to consider would be `maxUnavailable: 0` in a MCP?

Reply:
We're iterating on design with MCO, the original proposal to examine node specific details like kubelet version or kernel version may not be accepted in favor of something more abstract like the version of MachineConfig templates currently applied to nodes being greater than or equal to some version. At very high level we're just seeking to ensure that we enforce whatever host component version skew policy we come up with by setting Upgradeable=False to prevent upgrades that would violate those policies.

In 1) above MCO would only set Upgradeable=False when those constraints would be violated through a minor version upgrade, within this context we don't actually care if the pools are paused just that they're of some minimum version.

All operators are enabled to set Upgradeable=False, it's the primary mechanism used by operators to assert that the cluster is not in a state which would allow for minor version upgrades.

Reply:
@sdodson That is fine, for such information we have an RFE for the machine-os-content to provide such information via annotations as part of the extension system. Just trying to understand the high-level constraints and the best way to "guard" special resources in a cluster from unintended upgrades.

Reply:
The annotations are needed to support https://github.com/openshift/enhancements/pull/357 https://issues.redhat.com/browse/GRPA-2713

Reply:
`Upgradeable=False` How and where is this exactly set?

Reply:
@sdodson By all operators you mean operators that are managed via CVO and OLM? I had a question on my enhancement if OLM managed operators are allowed to create a ClusterOperator object to "control" upgradability.

Reply:
>  I had a question on my enhancement if OLM managed operators are allowed to create a ClusterOperator object to "control" upgradability.

I didn't think this was possible before, but turns out I was wrong, and the CVO [uses an unfiltered list of all ClusterOperator objects][1], regardless of whether they are from CVO manifests or added by third parties, when it calculates whether minor bumps are blocked.

[1]: https://github.com/openshift/cluster-version-operator/blob/bac69e03599f8c57f46585246618e462c31c2e93/pkg/cvo/upgradeable.go#L145

Reply:
@wking Thanks for pointing this out, started to read the code as well and doing some tests with custom MCPs and custom ClusterOperator objects.